xayn-ai/src/ranker/context.rs

```diff
@@ -1,119 +1,231 @@
+    NotEnoughPositiveCois { expected: u32, has: u32 },
+    /// Not enough negative cois (expected {expected:?}, has {has:?})
+    NotEnoughNegativeCois { expected: u32, has: u32 },
+    /// Failed to find the closest cois
+    FailedToFindTheClosestCois,
 }
 
-impl Context {
-    pub(crate) fn from_cois(cois: &[(DocumentId, CoiComponent)]) -> Self {
-        let cois_len = cois.len() as f32;
-        let pos_avg = cois.iter().map(|(_, coi)| coi.pos_distance).sum::<f32>() / cois_len;
-        let neg_max = cois
-            .iter()
-            .map(|(_, coi)| coi.neg_distance)
-            .fold(f32::MIN, f32::max); // NOTE f32::max considers NaN as smallest value
+/// Helper struct for [`find_closest_cois`].
+struct ClosestCois {
+    /// The ID of the positive centre of interest
+    pos_id: CoiId,
+    /// Distance from the positive centre of interest
+    pos_distance: f32,
+    pos_last_view: SystemTime,
+
+    /// Distance from the negative centre of interest
+    neg_distance: f32,
+    neg_last_view: SystemTime,
+}
 
-        Self { pos_avg, neg_max }
+fn find_closest_cois(
+    embedding: &Embedding,
+    user_interests: &UserInterests,
+    neighbors: usize,
+) -> Option<ClosestCois> {
+    let (pos_coi, pos_distance) = find_closest_coi(&user_interests.positive, embedding, neighbors)?;
+    let (neg_coi, neg_distance) = find_closest_coi(&user_interests.negative, embedding, neighbors)?;
+
+    ClosestCois {
+        pos_id: pos_coi.id(),
+        pos_distance,
+        pos_last_view: pos_coi.stats.last_view,
+        neg_distance,
+        neg_last_view: neg_coi.last_view,
     }
+    .into()
+}
+
+fn compute_score_for_embedding(
+    embedding: &Embedding,
+    user_interests: &UserInterests,
+    relevances: &mut RelevanceMap,
+    neighbors: usize,
+    horizon: Duration,
+    now: SystemTime,
+) -> Result<f32, Error> {
+    let cois = find_closest_cois(embedding, user_interests, neighbors)
+        .ok_or(Error::FailedToFindTheClosestCois)?;
+
+    let pos_decay = compute_coi_decay_factor(horizon, now, cois.pos_last_view);
+    let neg_decay = compute_coi_decay_factor(horizon, now, cois.neg_last_view);
+
+    relevances.compute_relevances(&user_interests.positive, horizon);
+    let pos_coi_relevance = relevances
+        .relevance_for_coi(&cois.pos_id).unwrap(/* we calculate relevance for all positive cois one line above */);
 
-    /// Calculates score from given positive distance and negative distance.
-    /// Both positive and negative distance must be >= 0.
-    pub(crate) fn calculate_score(&self, pos: f32, neg: f32) -> f32 {
-        debug_assert!(pos >= 0. && neg >= 0.);
-        let frac_pos = (self.pos_avg > 0.)
-            .then(|| (1. + pos / self.pos_avg).recip())
-            .unwrap_or(1.);
-        let frac_neg = (1. + (self.neg_max - neg)).recip();
+    let score = cois.pos_distance * pos_decay + pos_coi_relevance - cois.neg_distance * neg_decay;
```

Comment:
we could just return it without an additional binding
```suggestion
    Ok(cois.pos_distance * pos_decay + pos_coi_relevance - cois.neg_distance * neg_decay)
```